docs/rfcs/rfc-034-aggregation-query-filters.md

```diff
@@ -0,0 +1,257 @@
+# Aggregation Query Filters
+
+## Problem
+
+Given the following type definition:
+
+```graphql
+type Group {
+    name: String!
+    posts: [Post!]! @relationship(type: "HAS_POST", direction: OUT)
+}
+
+type Post {
+    title: String!
+    released: Int
+    author: Person @relationship(type: "WROTE_POST", direction: IN)
+}
+
+type Person {
+    name: String!
+    role: String!
+    posts: [Post!]! @relationship(type: "WROTE_POST", direction: OUT)
+}
+```
+
+Is currently possible to query all the groups that contain less than 10 `Post`, using a query like:
+
+```graphql
+query {
+  groups(where: {
+    postsAggregate: {
+      count_LT: 10
+    }
+  }) {
+    name
+  }
+}
+```
+
+**Cypher**
+
+```cypher
+MATCH (this:`Group`)
+CALL {
+    WITH this
+    MATCH (this:`Group`)-[this0:HAS_POST]->(this1:`Post`)
+    RETURN count(this1) < 10 AS var2
+}
+WITH *
+WHERE var2 = true
+RETURN this { .name } AS this
+```
+
+Although, is not possible to query all the groups that contain less than 10 `Posts` written in 2022.
+This is not possible because there is no such mechanism implemented that allows the user to modify the aggregation query.
+
+The Cypher required to achieve the above looks like this:
+
+**Cypher**
+```cypher
+MATCH (this:`Group`)
+CALL {
+    WITH this
+    MATCH (this:`Group`)-[this0:HAS_POST]->(this1:`Post`)
+    WHERE this1.released = 2022
+    RETURN count(this1) < 10 AS var2
+}
+WITH *
+WHERE var2 = true
+RETURN this { .name } AS this
+```
+
+## Proposed Solution
+
+This RFC propose to introduce the type: `{type}{relationField}AggregateFilter` and the `{type}WhereNoNested` as follow:
+
+```graphql
+input PostWhereNoNested {
+  OR: [PostWhereNoNested!]
+  AND: [PostWhereNoNested!]
+  title: String
+  title_NOT: String
+  title_IN: [String!]
+  title_NOT_IN: [String!]
+  title_CONTAINS: String
+  title_NOT_CONTAINS: String
+  title_STARTS_WITH: String
+  title_NOT_STARTS_WITH: String
+  title_ENDS_WITH: String
+  title_NOT_ENDS_WITH: String
+  released: Int
+  released_NOT: Int
+  released_IN: [Int]
+  released_NOT_IN: [Int]
+  released_LT: Int
+  released_LTE: Int
+  released_GT: Int
+  released_GTE: Int
+}
+
+input GroupPostsAggregateFilter {
```

Comment:
In case we choose to implement this, it will be good to find a better name rather than `queryWhere` and `aggregationWhere`!

Reply:
Yeah this was my main thought reading this.

I was just looking at how the problem has been solved elsewhere and it looks like mongo uses `$match` for the equivalent of `queryWhere` and `$group` for `aggregationWhere` which I think makes the distinction a bit clearer.

If I was to apply similar thinking to our situation I'd change `queryWhere` to `aggregateWhere` and  `aggregationWhere` to `group`.

What do you think?

Reply:
I think that in mongo you filter the Result set with `$match` and then you have the aggregation operation with `$group`, in our case what actually is not an aggregation, but a filter based on aggregation.
I'm happy with something like `aggregateWhere` and `groupWhere`, but, in my opinion, I believe that is better to wait for the API uniformity piece before taking a decision on this!

Reply:
I do think the naming of this is going to be tricky to get right because it's a weird concept to have a filter on a filter. To me having `where` in both the names makes it very difficult to distinguish between the behaviours.

I know the `group` input would be used as a filter but I don't think that means it needs to have `where` in the name in the same way we don't call other filters `titleWhere`, `releasedWhere`, `postsAggregateWhere` etc.

I'm happy to wait for the API uniformity work if you think it'll make the naming easier as you know more about what is happening there!